ts/screens/onboarding/OnboardingServicesPreferenceScreen.tsx

```diff
@@ -1,179 +1,214 @@
+    profileServicePreferencesModeSelector
+  );
+
   // if the user is not new and he/she hasn't a preference set, pre-set with AUTO mode
-  const mode = props.profileServicePreferenceMode;
-  const [modeSelected, setModeSelected] = React.useState<
+  const mode = profileServicePreferenceMode;
+  const [modeSelected, setModeSelected] = useState<
     ServicesPreferencesModeEnum | undefined
   >(mode);
-  const [prevPotProfile, setPrevPotProfile] = React.useState<
-    typeof props.potProfile
-  >(props.potProfile);
 
-  const { profileServicePreferenceMode, potProfile, onContinue } = props;
+  const dispatchServicesOptinCompleted = useCallback(
+    () => dispatch(servicesOptinCompleted()),
+    [dispatch]
+  );
 
-  useOnFirstRender(() => {
-    trackServiceConfigurationScreen(getFlowType(true, isFirstOnboarding));
-  });
+  const onServicePreferenceSelected = useCallback(
+    (mode: ServicesPreferencesModeEnum) =>
+      dispatch(
+        profileUpsert.request({ service_preferences_settings: { mode } })
+      ),
+    [dispatch]
+  );
 
-  const store = useStore();
+  const navigateToOnboardingServicePreferenceComplete = useCallback(() => {
+    navigation.navigate(ROUTES.ONBOARDING, {
+      screen: ROUTES.ONBOARDING_SERVICES_PREFERENCE_COMPLETE
+    });
+  }, [navigation]);
 
-  React.useEffect(() => {
-    // when the user made a choice (the profile is right updated), continue to the next step
-    if (isServicesPreferenceModeSet(profileServicePreferenceMode)) {
-      void trackServiceConfiguration(
-        profileServicePreferenceMode,
-        getFlowType(true, isFirstOnboarding),
-        store.getState()
-      );
-      onContinue(isFirstOnboarding);
-      return;
-    }
-    // show error toast only when the profile updating fails
-    // otherwise, if the profile is in error state, the toast will be shown immediately without any updates
-    if (!pot.isError(prevPotProfile) && pot.isError(potProfile)) {
-      showToast(I18n.t("global.genericError"));
-    }
-    setPrevPotProfile(potProfile);
-  }, [
-    isFirstOnboarding,
-    prevPotProfile,
-    potProfile,
-    profileServicePreferenceMode,
-    onContinue,
-    store
-  ]);
+  const onContinue = useCallback(
+    (isFirstOnboarding: boolean) =>
+      // if the user is not new, navigate to the thank-you screen
+      !isFirstOnboarding
+        ? navigateToOnboardingServicePreferenceComplete()
+        : dispatchServicesOptinCompleted(),
+    [
+      dispatchServicesOptinCompleted,
+      navigateToOnboardingServicePreferenceComplete
+    ]
+  );
 
   const handleOnContinue = () => {
-    if (modeSelected) {
-      props.onServicePreferenceSelected(modeSelected);
-    }
+    void trackServiceConfiguration(
+      profileServicePreferenceMode,
+      getFlowType(true, isFirstOnboarding),
+      store.getState()
+    );
+    onContinue(isFirstOnboarding);
+  };
+
+  const selectCurrentMode = (mode: ServicesPreferencesModeEnum) => {
```

Comment:
What about using a callback? What do you think?

Reply:
done!